File: src/connector.rs
use std::collections::BTreeMap;
use std::fmt::Display;
use std::io;
use std::path::Path;

use async_trait::async_trait;
use indexmap::IndexMap;
use serde_derive::{Deserialize, Serialize};
use serde_json::Map;
use url::Url;
use std::fs::read_to_string;

use ndc_hub::connector::{self, QueryError};
use ndc_hub::models::{self, RowFieldValue, MutationOperationResults, MutationOperation, Field};

#[derive(Clone, Default)]
pub struct TypescriptConnector;

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, schemars::JsonSchema)]
#[serde(untagged)]
pub enum TypescriptSource {
    Static(String),
    FromUrl { url: Url },
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, schemars::JsonSchema)]
pub struct RawConfiguration {
    pub typescript_source: Option<TypescriptSource>,
    pub schema_location: Option<String>, // Is there a better path type for this?
    pub deno_deployment_url: Option<String>,
}

impl Default for RawConfiguration {
    fn default() -> RawConfiguration {
        RawConfiguration {
            typescript_source: None,
            schema_location: None,
            deno_deployment_url: None,
        }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, schemars::JsonSchema)]
pub struct InputFunctionArgumentPositions {
    functions: Vec<InputFunctionInfo>,
    procedures: Vec<InputFunctionInfo>
}

impl InputFunctionArgumentPositions {
    fn fix(&self) -> FunctionArgumentPositions {

        let mut functions = IndexMap::new();

        let mut functions_and_procedures = self.functions.clone();
        let mut procedures_clone = self.procedures.clone();
        functions_and_procedures.append(&mut procedures_clone);

        for f in functions_and_procedures {

            let mut arguments = IndexMap::new();

            for (k, v) in f.arguments.iter() {
                arguments.insert(k.clone(), v.position);
            }

            functions.insert(f.name, ArgumentPositions { arguments });
        }

        FunctionArgumentPositions { functions }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, schemars::JsonSchema)]
pub struct InputFunctionInfo {
    name: String,
    arguments: IndexMap<String, InputArugmentInfo>
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, schemars::JsonSchema)]
pub struct InputArugmentInfo {
    position: u32
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, schemars::JsonSchema)]
pub struct ArgumentPositions {
    arguments: IndexMap<String, u32>
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, schemars::JsonSchema)]
pub struct FunctionArgumentPositions {
    functions: IndexMap<String, ArgumentPositions>
}

impl FunctionArgumentPositions {
    fn get(&self, func_name: &str, arg_name: &str) -> Option<u32> {
        if let Some(f) = self.functions.get(func_name) {
            if let Some(a) = f.arguments.get(arg_name) {
                return Some(*a);
            }
        }
        None
    }
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, schemars::JsonSchema)]
pub struct Configuration {
    pub deno_deployment_url: Url,
    pub schema: models::SchemaResponse,
    pub function_argument_positions: FunctionArgumentPositions,
}

#[derive(Clone, Debug)]
pub struct State {
    http_client: reqwest::Client,
}

#[async_trait]
impl connector::Connector for TypescriptConnector {
    type RawConfiguration = RawConfiguration;
    type Configuration = Configuration;
    type State = State;

    fn make_empty_configuration() -> Self::RawConfiguration {
        RawConfiguration::default()
    }

    async fn update_configuration(
        config: &Self::RawConfiguration,
    ) -> Result<Self::RawConfiguration, connector::UpdateConfigurationError> {
        Ok(config.clone())
    }

    async fn validate_raw_configuration(
        configuration: &Self::RawConfiguration,
    ) -> Result<Self::Configuration, connector::ValidateError> {

        // Why so many clones?
        let schema_location = configuration.schema_location.clone().unwrap_or("/schema.json".to_string());
        let schema_and_function_arguments = 
            read_schema(schema_location.clone())
                .map_err(|err| mk_single_validate_error("/schema.json", &format!("Couldn't read schema from {}: {}", schema_location.clone(), err)))?;

        let schema = schema_and_function_arguments.0;
        let function_argument_positions = schema_and_function_arguments.1;

        let deno_deployment_url_string = configuration.deno_deployment_url.clone().unwrap_or( "http://localhost:8000".to_string());

        let deno_deployment_url = Url::parse(&deno_deployment_url_string)
            .map_err(|_err| mk_single_validate_error("deno_deployment_url", "Couldn't parse deno deployment url."))?;

        Ok(Configuration {
            deno_deployment_url,
            schema,
            function_argument_positions
        })
    }

    async fn try_init_state(
        _configuration: &Self::Configuration,
        _metrics: &mut prometheus::Registry,
    ) -> Result<Self::State, connector::InitializationError> {
        Ok(State {
            http_client: reqwest::Client::new(),
        })
    }

    fn fetch_metrics(
        _configuration: &Self::Configuration,
        _state: &Self::State,
    ) -> Result<(), connector::FetchMetricsError> {
        todo!()
    }

    async fn health_check(
        _configuration: &Self::Configuration,
        _state: &Self::State,
    ) -> Result<(), connector::HealthError> {
        todo!()
    }

    async fn get_capabilities() -> models::CapabilitiesResponse {
        models::CapabilitiesResponse {
            versions: "^0.1.0".into(),
            capabilities: models::Capabilities {
                query: None,
                explain: None,
                mutations: None,
                relationships: None,
            },
        }
    }

    async fn get_schema(
        configuration: &Self::Configuration,
    ) -> Result<models::SchemaResponse, connector::SchemaError> {
        Ok(configuration.schema.clone())
    }

    async fn explain(
        _configuration: &Self::Configuration,
        _state: &Self::State,
        _request: models::QueryRequest,
    ) -> Result<models::ExplainResponse, connector::ExplainError> {
        todo!()
    }

    async fn mutation(
        configuration: &Self::Configuration,
        state: &Self::State,
        request: models::MutationRequest,
    ) -> Result<models::MutationResponse, connector::MutationError> {

        let mut operation_results: Vec<MutationOperationResults> = vec!();

        for op in request.operations {
            match op {
                MutationOperation::Procedure {name, arguments, fields} => {
                    let argument_literals = arguments.iter().map(|(k,v)| (k.clone(), models::Argument::Literal { value: v.clone()} )).collect();
                    let result = handle(configuration, state, name, argument_literals, fields).await
                        .map_err(query_error_to_mutation_error)?;
                    let mutation_operation_results = result_to_mutation_operation_results(result);
                    operation_results.push(mutation_operation_results);
                    ()
                }
                _ => {
                    () // TODO: Log unsupported types.
                }
            }
        }

        return Ok(models::MutationResponse { operation_results });
    }

    async fn query(
        configuration: &Self::Configuration,
        state: &Self::State,
        request: models::QueryRequest,
    ) -> Result<models::QueryResponse, connector::QueryError> {
        handle(configuration, state, request.collection, request.arguments, request.query.fields).await
    }
}

fn query_error_to_mutation_error(query_error: connector::QueryError) -> connector::MutationError {
    match query_error {
        QueryError::InvalidRequest(msg) => connector::MutationError::InvalidRequest(msg),
        QueryError::UnsupportedOperation(msg) => connector::MutationError::UnsupportedOperation(msg),
        QueryError::Other(err) => connector::MutationError::Other(err)
    }
}

fn result_to_mutation_operation_results(result: models::QueryResponse) -> MutationOperationResults {
    // TODO: Revisit this in order to respond to rows > 1
    for rs in result.0.iter() {
        return MutationOperationResults {
            affected_rows: 1,
            returning: Some(rs.rows.clone().unwrap_or(vec!()))
        }
    }
    return MutationOperationResults {
        affected_rows: 0,
        returning: None
    }
}

async fn handle(
    configuration: &Configuration,
    state: &State,
    collection: String,
    arguments: BTreeMap<String, models::Argument>,
    fields: Option<IndexMap<String, Field>> 
) -> Result<models::QueryResponse, connector::QueryError> {
    let function_name = collection;

    // Note: The arg names that are passed in here are actually indexed by the name suffix
    let mut indexed_args = arguments
        .into_iter()
        .map(|(arg_name, argument)| eval_argument(arg_name, argument))
        .collect::<Result<Vec<_>,_>>()?;

    indexed_args.sort_by_key(|(argument_name, _)|
        configuration.function_argument_positions.get(&function_name, argument_name)
    );

    let args = indexed_args.into_iter().map(|(_, v)| v).collect();

    let request_body = FunctionInvocation {
        function_name: function_name.clone(),
        args,
    };

    let http_response = state
        .http_client
        .post(configuration.deno_deployment_url.clone())
        .json(&request_body)
        .send()
        .await
        .map_err(|err| QueryError::Other(Box::new(err)))?;

    if http_response.status().is_success() {
        let response = http_response
            .json::<serde_json::Value>()
            .await
            .map_err(|err| QueryError::Other(Box::new(err)))?;

        let processed_response = process_fields(response, fields);

        let rows: Vec<IndexMap<String, RowFieldValue>> =
            vec![IndexMap::from_iter([(
                "__value".into(),
                RowFieldValue(processed_response)
            )])];

        Ok(models::QueryResponse(vec![models::RowSet {
            aggregates: None,
            rows: Some(rows),
        }]))

    } else {
        println!("There was an error executing the function: {}", function_name);

        let response = http_response
            .text()
            .await
            .map_err(|err| QueryError::Other(Box::new(err)))?;

        Err(QueryError::InvalidRequest(response))
    }

}

fn process_fields(response: serde_json::Value, fields: Option<IndexMap<String, Field>>) -> serde_json::Value {
    match fields {
        None => response,
        Some(fs) => {
            match response {
                serde_json::Value::Object(o) => {
                    let obj = fs.into_iter().filter_map(|(k,v)| {
                        match v {
                            Field::Column { column } => {
                                match o.get(&column) {
                                    None => Some((k, serde_json::Value::Null)),
                                    Some(v) => Some((k,v.clone()))
                                }
                            },
                            _ => None
                        }
                    }).collect::<Map<String, serde_json::Value>>();
                    serde_json::Value::Object(obj)
                },
                serde_json::Value::Array(a) => {
                    let arr = a.into_iter().map(|e| {
                        process_fields(e.clone(), Some(fs.clone()))
                    }).collect();
                    serde_json::Value::Array(arr)
                },
                _ => response
            }
        }
    }
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct FunctionInvocation {
    function_name: String,
    args: Vec<serde_json::Value>,
}

fn eval_argument(arg_name: String, arg: models::Argument) -> Result<(String, serde_json::Value), QueryError> {
    match arg {
        models::Argument::Literal { value } => Ok((arg_name, value)),
        models::Argument::Variable { .. } => Err(QueryError::UnsupportedOperation("Variables in arguments not supported".to_owned())),
    }
}

fn mk_single_validate_error(key: &str, message: &str) -> connector::ValidateError {
    let errs = vec![connector::InvalidRange {
        path: vec![connector::KeyOrIndex::Key(String::from(key))],
        message: String::from(message),
    }];
    connector::ValidateError::ValidateError(errs)
}

fn read_schema<P: AsRef<Path> + Display + Clone>(
    path: P
) -> Result<(models::SchemaResponse, FunctionArgumentPositions), Box<std::io::Error>> {
    
    let file_contents = read_to_string(path.clone())
        .map_err(|err| {
            // TODO: Format the path into the message rather than hardcoding the file
            let message = format!("Could not open {}: {}", path, err);
            let err = io::Error::new(io::ErrorKind::Other, message);
            Box::new(err)
        })?;

    let schema: models::SchemaResponse = serde_json::from_str(&file_contents)
        .map_err(|err| {
            let message = format!("functions.ts.schema.json should be valid SchemaResponse: {}", err);
            let err = io::Error::new(io::ErrorKind::Other, message);
            Box::new(err)
        })?;
    
    let positions_vector: InputFunctionArgumentPositions = serde_json::from_str(&file_contents)
        .map_err(|err| {
            let message = format!("functions.ts.schema.json should be valid FunctionArgumentPositions: {}", err);
            let err = io::Error::new(io::ErrorKind::Other, message);
            Box::new(err)
        })?;

    Ok((schema, positions_vector.fix()))
}

File: src/connector.ts
/**
 * Implementation of the Connector interface for Deno connector.
 * Using https://github.com/hasura/ndc-qdrant/blob/main/src/index.ts as an example.
 */

import { FunctionDefinitions, get_ndc_schema, NullOrUndefinability, ObjectTypeDefinitions, ProgramSchema, TypeDefinition } from "./schema.ts";
import { inferProgramSchema, Struct, } from "./infer.ts";
import { resolve } from "https://deno.land/std@0.208.0/path/mod.ts";
import { JSONSchemaObject } from "npm:@json-schema-tools/meta-schema";
import { isArray, unreachable } from "./util.ts";

import * as sdk from 'npm:@hasura/ndc-sdk-typescript@1.2.5';
export * as sdk from 'npm:@hasura/ndc-sdk-typescript@1.2.5';

export type State = {
  functions: RuntimeFunctions
}

export type RuntimeFunctions = {
  // deno-lint-ignore ban-types
  [function_name: string]: Function
}

export interface RawConfiguration {
  functions: string,
  port?: number, // Included only for punning Connector.start()
  hostname?: string, // Included only for punning Connector.start()
  schemaMode?: 'READ' | 'INFER',
  schemaLocation?: string,
  vendor?: string,
  preVendor?: boolean,
}

export const RAW_CONFIGURATION_SCHEMA: JSONSchemaObject = {
  description: 'Typescript (Deno) Connector Configuration',
  type: 'object',
  required: [ 'functions' ],
  properties:
  {
    functions: {
      description: 'Location of your functions entrypoint (default: ./functions/index.ts)',
      type: 'string'
    },
    vendor: {
      description: 'Location of dependencies vendor folder (optional)',
      type: 'string'
    },
    preVendor: {
      description: 'Perform vendoring prior to inference in a sub-process (default: true)',
      type: 'boolean'
    },
    schemaMode:  {
      description: 'INFER the schema from your functions, or READ it from a file.',
      type: "string",
      enum: ["READ", "INFER"]
    },
    schemaLocation: {
      description: 'Location of your schema file. schemaMode=READ reads the file, schemaMode=INFER writes the file (optional)',
      type: 'string'
    },
  }
};

type Configuration = {
  inferenceConfig: InferenceConfig,
  programSchema: ProgramSchema,
}

type InferenceConfig = {
  functions: string,
  schemaMode: 'READ' | 'INFER',
  schemaLocation?: string,
  vendorDir: string,
  preVendor: boolean,
}

export const CAPABILITIES_RESPONSE: sdk.CapabilitiesResponse = {
  versions: "^0.1.0",
  capabilities: {
    query: {
      variables: {}
    },
  },
};

type Payload = {
  function: string,
  args: Struct<unknown>
}


///////////////////// Helper Functions /////////////////////


/**
 * Performs analysis on the supplied program.
 * Expects that if there are dependencies then they will have been vendored.
 *
 * @param cmdObj
 * @returns Schema and argument position information
 */
export function getProgramSchema(cmdObj: InferenceConfig): ProgramSchema {
  switch(cmdObj.schemaMode) {
    /**
     * The READ option is available in case the user wants to pre-cache their schema during development.
     */
    case 'READ': {
      if(!cmdObj.schemaLocation) {
        throw new Error('--schema-location is required if using --schema-mode READ');
      }
      console.error(`Reading existing schema: ${cmdObj.schemaLocation}`);
      const bytes = Deno.readFileSync(cmdObj.schemaLocation);
      const decoder = new TextDecoder("utf-8");
      const decoded = decoder.decode(bytes);
      return JSON.parse(decoded);
    }
    case 'INFER': {
      console.error(`Inferring schema with map location ${cmdObj.vendorDir}`);
      const programSchema = inferProgramSchema(cmdObj.functions, cmdObj.vendorDir, cmdObj.preVendor);
      const schemaLocation = cmdObj.schemaLocation;
      if(schemaLocation) {
        console.error(`Writing schema to ${cmdObj.schemaLocation}`);
        const infoString = JSON.stringify(programSchema);
        // NOTE: Using sync functions should be ok since they're run on startup.
        Deno.writeTextFileSync(schemaLocation, infoString);
      }
      return programSchema;
    }
    default:
      throw new Error('Invalid schema-mode. Use READ or INFER.');
  }
}

/**
 * Performs invocation of the requested function.
 * Assembles the arguments into the correct order.
 * This doesn't catch any exceptions.
 *
 * @param functions
 * @param function_definitions
 * @param payload
 * @returns the result of invocation with no wrapper
 */
async function invoke(function_name: string, args: Record<string, unknown>, functions: RuntimeFunctions, program_schema: ProgramSchema): Promise<unknown> {
  const func = functions[function_name];
  const prepared_args = prepare_arguments(function_name, args, program_schema.functions, program_schema.object_types);

  try {
    const result = func.apply(undefined, prepared_args);
    // Resolve the result if it is a promise
    if (typeof result === "object" && 'then' in result && typeof result.then === "function") {
      return await result;
    }
    return result;
  } catch (e) {
    throw new sdk.InternalServerError(`Error encountered when invoking function ${function_name}`, { message: e.message, stack: e.stack });
  }
}

/**
 * This takes argument position information and a payload of function
 * and named arguments and returns the correctly ordered arguments ready to be applied.
 *
 * @param function_definitions
 * @param payload
 * @returns An array of the function's arguments in the definition order
 */
export function prepare_arguments(function_name: string, args: Record<string, unknown>, function_definitions: FunctionDefinitions, object_type_definitions: ObjectTypeDefinitions): unknown[] {
  const function_definition = function_definitions[function_name];

  if(!function_definition) {
    throw new sdk.InternalServerError(`Couldn't find function ${function_name} in schema.`);
  }

  return function_definition.arguments
    .map(argDef => coerce_argument_value(args[argDef.argument_name], argDef.type, [argDef.argument_name], object_type_definitions));
}

function coerce_argument_value(value: unknown, type: TypeDefinition, value_path: string[], object_type_definitions: ObjectTypeDefinitions): unknown {
  switch (type.type) {
    case "array":
      if (!isArray(value))
        throw new sdk.BadRequest(`Unexpected value in function arguments. Expected an array at '${value_path.join(".")}'.`);
      return value.map((element, index) => coerce_argument_value(element, type.element_type, [...value_path, `[${index}]`], object_type_definitions))

    case "nullable":
      if (value === null) {
        return type.null_or_undefinability == NullOrUndefinability.AcceptsUndefinedOnly
          ? undefined
          : null;
      } else if (value === undefined) {
        return type.null_or_undefinability == NullOrUndefinability.AcceptsNullOnly
          ? null
          : undefined;
      } else {
        return coerce_argument_value(value, type.underlying_type, value_path, object_type_definitions)
      }
    case "named":
      if (type.kind === "scalar") {
        // Scalars are currently treated as opaque values, which is a bit dodgy
        return value;
      } else {
        const object_type_definition = object_type_definitions[type.name];
        if (!object_type_definition)
          throw new sdk.InternalServerError(`Couldn't find object type '${type.name}' in the schema`);
        if (value === null || typeof value !== "object") {
          throw new sdk.BadRequest(`Unexpected value in function arguments. Expected an object at '${value_path.join(".")}'.`);
        }
        return Object.fromEntries(object_type_definition.properties.map(property_definition => {
          const prop_value = (value as Record<string, unknown>)[property_definition.property_name];
          return [property_definition.property_name, coerce_argument_value(prop_value, property_definition.type, [...value_path, property_definition.property_name], object_type_definitions)]
        }));
      }
    default:
      return unreachable(type["type"]);
  }
}

// TODO: https://github.com/hasura/ndc-typescript-deno/issues/26 Do deeper field recursion once that's available
function pruneFields(func: string, fields: Struct<sdk.Field> | null | undefined, result: unknown): unknown {
  // This seems like a bug to request {} fields when expecting a scalar response...
  // File with engine?
  if(!fields || Object.keys(fields).length == 0) {
    // TODO: https://github.com/hasura/ndc-typescript-deno/issues/21 How to log with SDK?
    console.error(`Warning: No fields present in query for function ${func}.`);
    return result;
  }

  const response: Struct<unknown> = {};

  if (result === null || Array.isArray(result) || typeof result !== "object") {
    throw new sdk.InternalServerError(`Function '${func}' did not return an object when expected to`);
  }

  for(const [k,v] of Object.entries(fields)) {
    switch(v.type) {
      case 'column':
        response[k] = (result as Record<string, unknown>)[v.column] ?? null; // Coalesce undefined into null to ensure we always have a value for a requested column
        break;
      default:
        console.error(`Function ${func} field of type ${v.type} is not supported.`);
    }
  }

  return response;
}

async function query(
  configuration: Configuration,
  state: State,
  functionName: string,
  requestArgs: Struct<unknown>,
  requestFields: { [k: string]: sdk.Field; } | null,
): Promise<unknown> {
  const result = await invoke(functionName, requestArgs, state.functions, configuration.programSchema);
  return pruneFields(functionName, requestFields, result);
}

function resolveArguments(
  func: string,
  requestArgs: Struct<sdk.Argument>,
  requestVariables: { [k: string]: unknown },
): Struct<unknown> {
  const args = Object.fromEntries(Object.entries(requestArgs).map(([k,v], _i) => {
    const t = v.type;
    switch(t) {
      case 'literal':
        return [k, v.value];
      case 'variable': {
        if(!(v.name in requestVariables)) {
          throw new Error(`Variable ${v.name} not found for function ${func}`);
        }
        const value = requestVariables[v.name];
        return [k, value];
      }
      default:
        return unreachable(t)
    }
  }));
  return args;
}

/**
 * See https://github.com/hasura/ndc-sdk-typescript for information on these interfaces.
 */
export const connector: sdk.Connector<RawConfiguration, Configuration, State> = {
  async try_init_state(
      config: Configuration,
      _metrics: unknown
  ): Promise<State> {
    const functionsArg = config.inferenceConfig.functions;
    const functionsURL = `file://${functionsArg}`; // NOTE: This is required to run directly from deno.land.
    const functions = await import(functionsURL);
    return {
      functions
    }
  },

  get_capabilities(_: Configuration): sdk.CapabilitiesResponse {
    return CAPABILITIES_RESPONSE;
  },

  get_raw_configuration_schema(): JSONSchemaObject {
    return RAW_CONFIGURATION_SCHEMA;
  },

  make_empty_configuration(): RawConfiguration {
    const conf: RawConfiguration = {
      functions: './functions/index.ts',
      vendor: './vendor'
    };
    return conf;
  },

  // TODO: https://github.com/hasura/ndc-typescript-deno/issues/27 Make this add in the defaults
  update_configuration(configuration: RawConfiguration): Promise<RawConfiguration> {
    return Promise.resolve(configuration);
  },

  validate_raw_configuration(configuration: RawConfiguration): Promise<Configuration> {
    if (configuration.functions.trim() === "") {
      throw new sdk.BadRequest("'functions' must be set to the location of the TypeScript file that contains your functions")
    }
    if (configuration.schemaMode === "READ" && !configuration.schemaLocation) {
      throw new sdk.BadRequest("'schemaLocation' must be set if 'schemaMode' is READ");
    }
    const inferenceConfig: InferenceConfig = {
      functions: resolve(configuration.functions),
      schemaMode: configuration.schemaMode ?? "INFER",
      preVendor: configuration.preVendor ?? true,
      schemaLocation: configuration.schemaLocation,
      vendorDir: resolve(configuration.vendor || "./vendor"),
    };
    const programSchema = getProgramSchema(inferenceConfig);
    return Promise.resolve({
      inferenceConfig,
      programSchema
    });
  },

  get_schema(config: Configuration): Promise<sdk.SchemaResponse> {
    return Promise.resolve(get_ndc_schema(config.programSchema));
  },

  // TODO: https://github.com/hasura/ndc-typescript-deno/issues/28 What do we want explain to do in this scenario?
  explain(
    _configuration: Configuration,
    _: State,
    _request: sdk.QueryRequest
  ): Promise<sdk.ExplainResponse> {
    throw new Error('Implementation of `explain` pending.');
  },

  // NOTE: query and mutation both make all functions available and discrimination is performed by the schema
  async query(
    configuration: Configuration,
    state: State,
    request: sdk.QueryRequest
  ): Promise<sdk.QueryResponse> {

    const rows = [];
    for(const variables of request.variables ?? [{}]) {
      const args = resolveArguments(request.collection, request.arguments, variables);
      const result = await query(configuration, state, request.collection, args, request.query.fields ?? null);
      rows.push({ '__value': result });
    }

    return [{
      aggregates: {},
      rows
    }];
  },

  async mutation(
    configuration: Configuration,
    state: State,
    request: sdk.MutationRequest
  ): Promise<sdk.MutationResponse> {
    const results: Array<sdk.MutationOperationResults> = [];
    for(const op of request.operations) {
      switch(op.type) {
        case 'procedure': {
          const result = await query(configuration, state, op.name, op.arguments, op.fields ?? null);
          results.push({
            affected_rows: 1,
            returning: [{
              '__value': result
            }]
          });
          break;
        }
        default:
          throw new Error(`Mutation type ${op.type} not supported.`);
      }
    }
    return {
      operation_results: results
    }
  },

  // If the connector starts successfully it should be healthy
  health_check(_: Configuration, __: State): Promise<undefined> {
    return Promise.resolve(undefined);
  },

  // TODO: https://github.com/hasura/ndc-typescript-deno/issues/29 https://qdrant.github.io/qdrant/redoc/index.html#tag/service/operation/metrics
  fetch_metrics(_: Configuration, __: State): Promise<undefined> {
    return Promise.resolve(undefined);
  },
};
